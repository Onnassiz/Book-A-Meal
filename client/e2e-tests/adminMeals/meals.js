/* eslint func-names: 0, no-unused-expressions: 0 */
const path = require('path');

const imagePath = path.join(__dirname, '../../assets/images/3.jpg');

const meals = {
  'Should sign user into the application': (client) => {
    client
      .url('http://localhost:3002/#/')
      .waitForElementVisible('header')
      .moveToElement('#login', 10, 10)
      .click('#login')
      .waitForElementVisible('h3')
      .assert.containsText('h3', 'Sign In')
      .setValue('#signInEmail', 'onnassiz@gmail.com')
      .setValue('#signInPassword', 'password')
      .click('input[type=submit]')
      .waitForElementVisible('header')
      .assert.visible('#caterer_dropdown')
      .assert.visible('#user_dropdown');
  },

  'Should navigate to the admin meals page': (client) => {
    client
      .moveToElement('#caterer_dropdown', 10, 10)
      .pause(1000)
      .waitForElementVisible('#admin_meals')
      .assert.visible('#admin_meals')
      .click('#admin_meals')
      .waitForElementVisible('#addMeal')
      .pause(2000)
      .assert.visible('#addMeal');
  },

  'Should show new meal modal form': (client) => {
    client
      .moveToElement('#addMeal', 10, 10)
      .pause(1000)
      .assert.containsText('#addMeal', 'Add Meal')
      .click('#addMeal')
      .pause(1000)
      .waitForElementVisible('.box')
      .assert.visible('.box')
      .pause(2000);
  },

  'Should fill meal form and create a new meal': (client) => {
    client
      .pause(1000)
      .setValue('#name', 'Fried Rice')
      .setValue('#price', 2000)
      .setValue('#category', 'Dinner')
      .setValue('#description', 'This is a good meal. It has all you need. Try it out.')
      .moveToElement('input[type=submit]', 10, 10)
      .pause(2000)
      .click('input[type=submit]')
      .waitForElementVisible('div.Toastify__toast.Toastify__toast--default', 5000)
      .assert.visible('div.Toastify__toast.Toastify__toast--default');
  },

  'Should show a new meal modal form again': (client) => {
    client
      .moveToElement('#addMeal', 10, 10)
      .assert.containsText('#addMeal', 'Add Meal')
      .click('#addMeal')
      .waitForElementVisible('.box')
      .assert.visible('.box')
      .pause(1000);
  },

  'Should fill another meal form and create a new meal': (client) => {
    client
      .pause(1000)
      .setValue('#name', 'Chicken')
      .setValue('#price', 2000)
      .setValue('#category', 'Dinner')
      .setValue('#description', 'This is a good meal. It has all you need. Try it out.')
      .moveToElement('input[type=submit]', 10, 10)
      .pause(2000)
      .click('input[type=submit]')
      .waitForElementVisible('div.Toastify__toast.Toastify__toast--default', 5000)
      .assert.visible('div.Toastify__toast.Toastify__toast--default');
  },

  'Should show a new meal modal form for the third time': (client) => {
    client
      .moveToElement('#addMeal', 10, 10)
      .assert.containsText('#addMeal', 'Add Meal')
      .click('#addMeal')
      .waitForElementVisible('.box')
      .assert.visible('.box')
      .pause(1000);
  },

  'Should fill the third meal form and create a new meal': (client) => {
    client
      .pause(1000)
      .setValue('#name', 'Beans')
      .setValue('#price', 1000)
      .setValue('#category', 'Dinner')
      .setValue('#description', 'This is a good meal for me and you.')
      .moveToElement('input[type=submit]', 10, 10)
      .pause(2000)
      .click('input[type=submit]')
      .waitForElementVisible('div.Toastify__toast.Toastify__toast--default', 5000)
      .assert.visible('div.Toastify__toast.Toastify__toast--default');
  },

  'Should show an update meal modal': (client) => {
    client
      .waitForElementVisible('#updateModal:first-of-type')
      .assert.visible('#updateModal:first-of-type')
      .click('#updateModal:first-of-type')
      .assert.visible('.box')
      .pause(1000);
  },

  'Should update the third meal form and update a new meal': (client) => {
    client
      .pause(1000)
      .clearValue('#name')
      .setValue('#name', 'Rice')
      .clearValue('#price')
      .setValue('#price', 1200)
      .moveToElement('input[type=submit]', 10, 10)
      .pause(1000)
      .click('input[type=submit]')
      .waitForElementVisible('div.Toastify__toast.Toastify__toast--default', 5000)
      .assert.visible('div.Toastify__toast.Toastify__toast--default');
  },

  'Should show delete modal and delete meal on button click': (client) => {
    client
      .assert.visible('#deleteModal:first-of-type')
      .click('#deleteModal:first-of-type')
      .waitForElementVisible('button.swal-button.swal-button--confirm.swal-button--danger', 10000)
      .pause(1000)
      .click('button.swal-button.swal-button--confirm.swal-button--danger')
      .waitForElementVisible('button.swal-button.swal-button--confirm', 15000)
      .pause(1000)
      .click('button.swal-button.swal-button--confirm')
      .waitForElementNotVisible('button.swal-button.swal-button--confirm', 20000)
      .assert.hidden('button.swal-button.swal-button--confirm')
      .pause(1000);
  },

  'Should show an image upload modal modal': (client) => {
    client
      .assert.visible('#addPhotoModal:first-of-type')
      .click('#addPhotoModal:first-of-type')
      .waitForElementVisible('#dropzoneHeader')
      .assert.containsText('#dropzoneHeader', 'Drop Image Here')
      .pause(1000)
      .moveToElement('#upload_image', 10, 10)
      .pause(1000)
      .setValue('input[type=file]', imagePath)
      .waitForElementVisible('div.Toastify__toast.Toastify__toast--default', 15000)
      .assert.visible('div.Toastify__toast.Toastify__toast--default')
      .pause(1000);
    client.end();
  },
};

export default meals;
