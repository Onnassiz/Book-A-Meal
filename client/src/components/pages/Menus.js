import React, { Component } from 'react';
import MenusContainer from '../container/MenusContainer';
import NavBarContainer from '../container/NavBarContainer';

class Menus extends Component {
	render() {
		return (
			<div>
				<NavBarContainer page="menus" />
				<MenusContainer />
			</div>
		);
	}
}

export default Menus;
