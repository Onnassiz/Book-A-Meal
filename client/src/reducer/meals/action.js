import axios from 'axios';
import { store } from '../../store';
import { setLoading, unsetLoading } from '../formState/action';

export const SET_MEAL = 'SET_MEAL';
export const SET_MEAL_SERVER_ERRORS = 'SET_MEAL_SERVER_ERRORS';
export const SET_MEAL_ALERT = 'SET_MEAL_ALERT';

function setAlert(alert) {
	return dispatch => dispatch({
		type: SET_MEAL_ALERT,
		alert,
	});
}

export function setMeal(meal) {
	let { meals } = store.getState().meals;
	meals = [meal].concat(meals);
	return dispatch => dispatch({
		type: SET_MEAL,
		meals,
	});
}

export function setMealArray(meal) {
	let { meals } = store.getState().meals;
	meals = meals.concat(meal);
	return dispatch => dispatch({
		type: SET_MEAL,
		meals,
	});
}

export function updateMealState(meal) {
	const { meals } = store.getState().meals;
	const index = meals.findIndex(x => x.id === meal.id);
	meals[index] = meal;
	return dispatch => dispatch({
		type: SET_MEAL,
		meals,
	});
}

export function updateMealAfterDelete(id) {
	let { meals } = store.getState().meals;
	meals = meals.filter(x => x.id !== id);
	return dispatch => dispatch({
		type: SET_MEAL,
		meals,
	});
}

function setMealErrors(errors) {
	return dispatch => dispatch({
		type: SET_MEAL_SERVER_ERRORS,
		errors,
	});
}

export function postMeal(meal) {
	return (dispatch) => {
		dispatch(setLoading());
		return axios.post('api/v1/meals', meal).then((response) => {
			dispatch(unsetLoading());
			dispatch(setMeal(response.data));
			dispatch(setAlert('Meal successfully created'));
			return response;
		}).catch((error) => {
			dispatch(unsetLoading());
			dispatch(setMealErrors(error.response));
			return error;
		});
	};
}

export function updateMeal(meal) {
	return (dispatch) => {
		dispatch(setLoading());
		return axios.put(`api/v1/meals/${meal.id}`, meal).then((response) => {
			dispatch(unsetLoading());
			dispatch(updateMealState(response.data));
			dispatch(setAlert('Meal successfully updated'));
			return response;
		}).catch((error) => {
			dispatch(unsetLoading());
			dispatch(setMealErrors(error.response.data));
			return error;
		});
	};
}

export function deleteMealById(id) {
	return (dispatch) => {
		dispatch(setLoading());
		return axios.delete(`api/v1/meals/${id}`).then((response) => {
			dispatch(unsetLoading());
			dispatch(updateMealAfterDelete(id));
			dispatch(setAlert('Meal successfully deleted'));
			return response;
		}).catch((error) => {
			dispatch(unsetLoading());
			dispatch(setMealErrors(error.response.data));
			return error;
		});
	};
}

export function putImage(id, field) {
	return (dispatch) => {
		dispatch(setLoading());
		return axios.put(`api/v1/meals/image/${id}`, field).then((response) => {
			dispatch(unsetLoading());
			dispatch(updateMealState(response.data));
			dispatch(setAlert('Image successfully uploaded'));
			return response;
		}).catch((error) => {
			dispatch(unsetLoading());
			dispatch(setMealErrors(error.response.data));
			return error;
		});
	};
}
export function getMeals() {
	return (dispatch) => {
		dispatch(setLoading());
		return axios.get('api/v1/meals').then((response) => {
			dispatch(unsetLoading());
			dispatch(setMealArray(response.data));
			return response;
		}).catch((error) => {
			dispatch(unsetLoading());
			return error;
		});
	};
}
