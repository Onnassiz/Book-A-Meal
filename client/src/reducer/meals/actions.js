import axios from 'axios';
import { setLoading, unsetLoading } from '../formState/actions';

export const SET_MEAL = 'SET_MEAL';
export const SET_MEAL_SERVER_ERRORS = 'SET_MEAL_SERVER_ERRORS';
export const SET_MEAL_ALERT = 'SET_MEAL_ALERT';

function setAlert(alert) {
  return dispatch => dispatch({
    type: SET_MEAL_ALERT,
    alert,
  });
}

// function setMeal(storeMeals, meal) {
//   const meals = [meal].concat(storeMeals);
//   return dispatch => dispatch({
//     type: SET_MEAL,
//     meals,
//   });
// }

// function setMealArray(storeMeals, meal) {
//   const meals = storeMeals.concat(meal);
//   return dispatch => dispatch({
//     type: SET_MEAL,
//     meals,
//   });
// }

export function updateMealState(storeMeals, meal) {
  const index = storeMeals.findIndex(x => x.id === meal.id);
  const meals = storeMeals;
  meals[index] = meal;
  return dispatch => dispatch({
    type: SET_MEAL,
    meals,
  });
}

export function updateMealAfterDelete(storeMeals, id) {
  const meals = storeMeals.filter(x => x.id !== id);
  return dispatch => dispatch({
    type: SET_MEAL,
    meals,
  });
}

function setMealErrors(errors) {
  return dispatch => dispatch({
    type: SET_MEAL_SERVER_ERRORS,
    errors,
  });
}

export function postMeal(meal) {
  return (dispatch) => {
    dispatch(setLoading());
    return axios.post('api/v1/meals', meal).then((response) => {
      dispatch(unsetLoading());
      // dispatch(setMeal(getState().meals.meals, response.data));
      dispatch(setAlert('Meal successfully created'));
      return response;
    }).catch((error) => {
      dispatch(unsetLoading());
      dispatch(setMealErrors(error.response.data));
      return error;
    });
  };
}

export function updateMeal(meal) {
  return (dispatch) => {
    dispatch(setLoading());
    return axios.put(`api/v1/meals/${meal.id}`, meal).then((response) => {
      dispatch(unsetLoading());
      // dispatch(updateMealState(getState().meals.meals, response.data));
      dispatch(setAlert('Meal successfully updated'));
      return response;
    }).catch((error) => {
      dispatch(unsetLoading());
      dispatch(setMealErrors(error.response.data));
      return error;
    });
  };
}

export function deleteMealById(id) {
  return (dispatch) => {
    dispatch(setLoading());
    return axios.delete(`api/v1/meals/${id}`).then((response) => {
      dispatch(unsetLoading());
      // dispatch(updateMealAfterDelete(getState().meals.meals, id));
      return response;
    }).catch((error) => {
      dispatch(unsetLoading());
      dispatch(setMealErrors(error.response.data));
      return error;
    });
  };
}

export function getMeals(limit = 10, offset = 0) {
  return (dispatch) => {
    dispatch(setLoading());
    return axios.get(`api/v1/meals/?limit=${limit}&offset=${offset}`).then((response) => {
      dispatch(unsetLoading());
      // dispatch(setMealArray(getState().meals.meals, response.data));
      return response;
    }).catch((error) => {
      dispatch(unsetLoading());
      return error;
    });
  };
}
