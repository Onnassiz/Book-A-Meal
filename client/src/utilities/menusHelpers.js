
/* eslint no-underscore-dangle: 0 */
import { convertUnixToDateForUpdate, getDateFromMoment } from './functions';

export function registerMethods($thisBind) {
  const $this = $thisBind;
  $this.onChange = $this.onChange.bind($this);
  $this.updateUnits = $this.updateUnits.bind($this);
  $this.addToCart = $this.addToCart.bind($this);
  $this.removeFromCart = $this.removeFromCart.bind($this);
  $this.showMore = $this.showMore.bind($this);
  $this.showCalender = $this.showCalender.bind($this);
  $this.showCart = $this.showCart.bind($this);
  $this.closeCartPane = $this.closeCartPane.bind($this);
}

export function getMenuForDate(number, $this) {
  const { getMealsInDailyMenu } = $this.props;
  let date = new Date($this.state.date);
  date = convertUnixToDateForUpdate(date.setDate(date.getDate() + number) / 1000);
  getMealsInDailyMenu(date).then((response) => {
    if (response.status === 200) {
      $this.setState({
        date, mealsCount: response.data.count, meals: response.data.meals, activePage: 1,
      });
    }
  });
}

export function handlePageChange(pageNumber, $this) {
  const { getMealsInDailyMenu } = $this.props;
  const offset = (pageNumber - 1) * 12;
  getMealsInDailyMenu($this.state.date, offset, 12).then((response) => {
    if (response.status === 200) {
      $this.setState({
        activePage: pageNumber,
        meals: response.data.meals,
        mealsCount: response.data.count,
      });
    }
  });
}

export function handleDateChange(date, $this) {
  const { getMealsInDailyMenu } = $this.props;
  const selectedDate = getDateFromMoment(date._d);
  getMealsInDailyMenu(selectedDate).then((response) => {
    if (response.status === 200) {
      $this.setState({
        date: selectedDate,
        mealsCount: response.data.count,
        meals: response.data.meals,
        activePage: 1,
      });
    }
  });
}
