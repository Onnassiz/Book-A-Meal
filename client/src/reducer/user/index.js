import { SET_USER, SET_SIGN_UP_SERVER_ERRORS, SET_SIGN_IN_SERVER_ERRORS } from './action';

const initialState = {};

export default (state = initialState, action) => {
	switch (action.type) {
	case SET_USER:
		return { name: action.name, email: action.email, role: action.role };
	case SET_SIGN_UP_SERVER_ERRORS:
		return { signUpErrors: action.errors };
	case SET_SIGN_IN_SERVER_ERRORS:
		return { signInErrors: action.errors };
	default:
		return state;
	}
};
