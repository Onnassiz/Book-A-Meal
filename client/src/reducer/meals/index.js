import { SET_MEAL, SET_MEAL_SERVER_ERRORS } from './actions';

const initialState = {
  meals: [],
};

export default (state = initialState, action) => {
  switch (action.type) {
    case SET_MEAL:
      return { ...state, meals: action.meals };
    case SET_MEAL_SERVER_ERRORS:
      return { ...state, errors: action.errors };
    default:
      return state;
  }
};
