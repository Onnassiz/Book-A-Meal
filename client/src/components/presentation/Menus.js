/* eslint no-underscore-dangle: 0 */
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import swal from 'sweetalert';
import Pagination from 'react-js-pagination';
import SlidingPane from 'react-sliding-pane';
import empty from 'is-empty';
import moment from 'moment';
import Calendar from 'rc-calendar';
import 'react-sliding-pane/dist/react-sliding-pane.css';
import 'rc-calendar/assets/index.css';
import { convertUnixToDateForUpdate } from '../../utilities/functions';
import { getMenuForDate, registerMethods, handlePageChange, handleDateChange } from '../../utilities/menusHelpers';
import Card from './partials/MealCard';

class Menus extends Component {
  constructor(props) {
    super(props);
    this.state = {
      mealsCount: 0,
      activePage: 1,
      isPaneOpenLeft: false,
      meals: [],
      date: props.menus.currentDate,
    };
    registerMethods(this);
  }

  componentDidMount() {
    document.title = 'Menus - Just Eat';
    const { getMealsInDailyMenu } = this.props;
    getMealsInDailyMenu(this.state.date).then((response) => {
      if (response.status === 200) {
        this.setState({ mealsCount: response.data.count, meals: response.data.meals });
      }
    });
  }

  onChange(e) {
    const { getMealsInDailyMenu } = this.props;
    let date = new Date(e.target.value);
    date = convertUnixToDateForUpdate(date.setDate(date.getDate()) / 1000);
    return getMealsInDailyMenu(date).then(() => {
      this.setState({ date });
    });
  }

  handleDateChange(date) {
    handleDateChange(date || moment(new Date()), this);
  }

  handlePageChange(number) {
    handlePageChange(number, this);
  }

  showMore(meal) {
    const details = document.createElement('div');
    details.className = 'meal-details';
    details.innerHTML = `<h2>Name</h2><p>${meal.name}</p><h2>Description</h2><p>${meal.description}</p><h2>Category</h2><p>${meal.category}</p><h2>Catering Service</h2><p>${meal.caterer}</p>`;
    swal({
      title: 'Details',
      content: details,
    });
  }

  showCalender() {
    this.setState({ isPaneOpenLeft: true });
  }

  addToCart(meal) {
    const { addToCart, user } = this.props;
    addToCart(meal, user.id);
  }

  removeFromCart(mealId) {
    const { deleteFromCart, user } = this.props;
    deleteFromCart(mealId, user.id);
  }

  renderCalender() {
    return (
      <SlidingPane
        isOpen={this.state.isPaneOpenLeft}
        title="Pick a Date"
        from="left"
        width="320px"
        onRequestClose={() => this.setState({ isPaneOpenLeft: false })}
      >
        <div>
          <Calendar
            showDateInput={false}
            value={moment(this.state.date)}
            onChange={date => this.handleDateChange(date)}
          />
        </div>
      </SlidingPane>
    );
  }

  renderPagination() {
    return (
      <div>{this.state.mealsCount > 12 ?
        <div style={{ textAlign: 'center' }}>
          <Pagination
            hideDisabled
            activePage={this.state.activePage}
            totalItemsCount={this.state.mealsCount}
            itemsCountPerPage={12}
            pageRangeDisplayed={5}
            onChange={number => this.handlePageChange(number)}
          />
        </div> : ''}
      </div>
    );
  }

  renderCard(meal) {
    const { cart, menus } = this.props;
    return (
      <Card
        showMore={() => this.showMore(meal)}
        cart={cart.cart}
        addToCart={() => this.addToCart(meal)}
        meal={meal}
        key={meal.id + meal.caterer}
        currentDate={menus.currentDate}
        removeFromCart={() => this.removeFromCart(meal.id)}
      />
    );
  }

  renderMeals() {
    return (
      <div className="col-12">
        <div className="dateInput">
          <a onClick={() => getMenuForDate(-1, this)} title="back"><i className="ion-ios-skipbackward" /> Previous</a>
          <a onClick={() => getMenuForDate(1, this)} title="next day">Next Day <i className="ion-ios-skipforward" /></a>
          <h2>{new Date(this.state.date).toDateString()}</h2>
        </div>
        <div id="card_container">
          {empty(this.state.meals) ?
            <div id="no-menu">
              <h2>No menu has been set for this day</h2>
            </div> : this.state.meals.map(meal => this.renderCard(meal))}
        </div>
      </div>
    );
  }

  renderFixedCalender() {
    return (
      <div>
        <div className="control">
          <i className="ion-ios-calendar-outline" onClick={this.showCalender} />
        </div>
      </div>
    );
  }

  render() {
    return (
      <div id="content-body">
        {this.renderFixedCalender()}
        {this.renderCalender()}
        {this.renderMeals()}
        {this.renderPagination()}
      </div>
    );
  }
}

Menus.propTypes = {
  menus: PropTypes.object.isRequired,
  user: PropTypes.object.isRequired,
  cart: PropTypes.object.isRequired,
  getMealsInDailyMenu: PropTypes.func.isRequired,
  addToCart: PropTypes.func.isRequired,
  deleteFromCart: PropTypes.func.isRequired,
};
export default Menus;

