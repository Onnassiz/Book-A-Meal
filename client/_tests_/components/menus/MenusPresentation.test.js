import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import Pagination from 'react-js-pagination';
import SlidingPane from 'react-sliding-pane';
import Calendar from 'rc-calendar';
import Card from './../../../src/components/presentation/partials/MealCard';
import Menus from '../../../src/components/presentation/Menus';
import user from '../../objectProps/user.props';
import * as menusHelpers from '../../../src/utilities/menusHelpers';
import { getMeals } from '../../promiseProps/meals.props';
import { singleMeal, getMealsMock } from '../../objectProps/meals.props';

const props = {
  formState: { isLoading: false },
  menus: { currentDate: '2018-09-09' },
  user,
  cart: {
    cart: getMealsMock(1),
    owner: '',
    orderId: '',
    address: '',
    contact: '',
    isOpen: false,
    totalPrice: 0,
    updateMode: false,
  },
  getMealsInDailyMenu: getMeals,
  addToCart: jest.fn(),
  deleteFromCart: jest.fn(),
};

const wrapper = shallow(<Menus {...props} />);

describe('<Menus />', () => {
  it('Should render the Menus container successfully', () => {
    wrapper.setState({ meals: getMealsMock(1) });
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('Should call onChange method to update handle date change', () => {
    const onChangeSpy = jest.spyOn(wrapper.instance(), 'onChange');
    const firstEvent = {
      target: {
        name: 'date',
        value: '2019-12-12',
      },
    };
    wrapper.instance().onChange(firstEvent).then(() => {
      expect(wrapper.state().date).toBe('2019-12-12');
      expect(onChangeSpy).toHaveBeenCalled();
    });
  });

  it('Should trigger showMore to show more information about a meal cart in a modal', () => {
    const showMoreSpy = jest.spyOn(wrapper.instance(), 'showMore');
    wrapper.instance().showMore(singleMeal);
    expect(showMoreSpy).toHaveBeenCalled();
  });

  it('Should show fixed calendar when button is clicked', () => {
    wrapper.find('.ion-ios-calendar-outline').simulate('click');
    expect(wrapper.state().isPaneOpenLeft).toBe(true);
  });

  it('Should trigger remove item from cart when remove button is clicked', () => {
    const removeFromCart = jest.spyOn(wrapper.instance(), 'removeFromCart');
    wrapper
      .find(Card)
      .first()
      .dive()
      .find('.add_button')
      .simulate('click');

    expect(removeFromCart).toHaveBeenCalled();
  });

  it('Should trigger showMore function when more-button button is clicked', () => {
    const showMoreSpy = jest.spyOn(wrapper.instance(), 'showMore');
    wrapper
      .find(Card)
      .first()
      .dive()
      .find('.more-button')
      .simulate('click');

    expect(showMoreSpy).toHaveBeenCalled();
  });

  it('Should trigger add item to cart when add button is clicked', () => {
    const addToCartSpy = jest.spyOn(wrapper.instance(), 'addToCart');
    wrapper.setProps({
      cart: {
        cart: [],
        owner: '',
        orderId: '',
        address: '',
        contact: '',
        isOpen: false,
        totalPrice: 0,
        updateMode: false,
      },
    });

    wrapper
      .find(Card)
      .first()
      .dive()
      .find('.add_button')
      .simulate('click');

    expect(addToCartSpy).toHaveBeenCalled();
  });

  it('Should render pagination when mealsCount is more than 12', () => {
    const renderPaginationSpy = jest.spyOn(wrapper.instance(), 'renderPagination');
    wrapper.setState({ mealsCount: 13 });

    expect(wrapper.find(Pagination).length).toBe(1);
    expect(renderPaginationSpy).toHaveBeenCalled();
  });

  it('Should call handlePageChange when pagination button is clicked', () => {
    const handlePageChangeSpy = jest.spyOn(wrapper.instance(), 'handlePageChange');

    wrapper.find(Pagination).first().find('a').simulate('change');
    expect(handlePageChangeSpy).toHaveBeenCalled();
  });

  it('Should call handleDateChange when calendar date button is clicked', () => {
    const handleDateChangeSpy = jest.spyOn(wrapper.instance(), 'handleDateChange');

    wrapper.find(Calendar).simulate('change');
    expect(handleDateChangeSpy).toHaveBeenCalled();
  });

  it('Should set isPaneOpenLeft to false when slide-pane__close is clicked', () => {
    const handleDateChangeSpy = jest.spyOn(wrapper.instance(), 'handleDateChange');

    wrapper.find(SlidingPane).simulate('requestClose');
    expect(handleDateChangeSpy).toHaveBeenCalled();
  });

  it('Should change date when next or prev buttons are clicked', () => {
    const getMenuForDateSpy = jest.spyOn(menusHelpers, 'getMenuForDate');

    wrapper.find('.dateInput').find('a').first().simulate('click');
    wrapper.find('.dateInput').find('a').last().simulate('click');
    expect(getMenuForDateSpy).toHaveBeenCalledTimes(2);
  });
});
