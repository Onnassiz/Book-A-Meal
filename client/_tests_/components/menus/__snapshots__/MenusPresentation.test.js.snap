// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Menus /> Should render the Menus container successfully 1`] = `
<div
  id="content-body"
>
  <div>
    <div
      className="control"
    >
      <i
        className="ion-ios-calendar-outline"
        onClick={[Function]}
      />
    </div>
  </div>
  <ReactSlidingPane
    from="left"
    isOpen={false}
    onRequestClose={[Function]}
    title="Pick a Date"
    width="320px"
  >
    <div>
      <Calendar
        className=""
        locale={
          Object {
            "backToToday": "Back to today",
            "clear": "Clear",
            "dateFormat": "M/D/YYYY",
            "dateSelect": "select date",
            "dateTimeFormat": "M/D/YYYY HH:mm:ss",
            "dayFormat": "D",
            "decadeSelect": "Choose a decade",
            "month": "Month",
            "monthBeforeYear": true,
            "monthSelect": "Choose a month",
            "nextCentury": "Next century",
            "nextDecade": "Next decade",
            "nextMonth": "Next month (PageDown)",
            "nextYear": "Next year (Control + right)",
            "now": "Now",
            "ok": "Ok",
            "previousCentury": "Last century",
            "previousDecade": "Last decade",
            "previousMonth": "Previous month (PageUp)",
            "previousYear": "Last year (Control + left)",
            "timeSelect": "select time",
            "today": "Today",
            "weekSelect": "Choose a week",
            "year": "Year",
            "yearFormat": "YYYY",
            "yearSelect": "Choose a year",
          }
        }
        onChange={[Function]}
        onClear={[Function]}
        onKeyDown={[Function]}
        onOk={[Function]}
        onPanelChange={[Function]}
        onSelect={[Function]}
        prefixCls="rc-calendar"
        renderFooter={[Function]}
        renderSidebar={[Function]}
        showDateInput={false}
        showToday={true}
        style={Object {}}
        timePicker={null}
        value={"2018-09-08T23:00:00.000Z"}
        visible={true}
      />
    </div>
  </ReactSlidingPane>
  <div
    className="col-12"
  >
    <div
      className="dateInput"
    >
      <a
        onClick={[Function]}
        title="back"
      >
        <i
          className="ion-ios-skipbackward"
        />
         Previous
      </a>
      <a
        onClick={[Function]}
        title="next day"
      >
        Next Day 
        <i
          className="ion-ios-skipforward"
        />
      </a>
      <h2>
        Sun Sep 09 2018
      </h2>
    </div>
    <div
      id="card_container"
    >
      <Card
        addToCart={[Function]}
        cart={
          Array [
            Object {
              "category": "dolor sit",
              "caterer": "Ben",
              "description": "Maecenas ut massa quis augue luctus tincidunt. Nulla mollis molestie lorem. Quisque ut erat. Curabitur gravida nisi at nibh.",
              "id": "84c6d80c-3d4a-4772-856c-0b3825bc5333",
              "imageUrl": "http://thisImage.jpeg",
              "name": "Malvaceae",
              "price": 400,
              "totalPrice": 2000,
              "units": 2,
              "userId": "2794fddd-14f3-4ec9-a1dc-88ad2bc649da",
            },
          ]
        }
        currentDate="2018-09-09"
        key="84c6d80c-3d4a-4772-856c-0b3825bc5333Ben"
        meal={
          Object {
            "category": "dolor sit",
            "caterer": "Ben",
            "description": "Maecenas ut massa quis augue luctus tincidunt. Nulla mollis molestie lorem. Quisque ut erat. Curabitur gravida nisi at nibh.",
            "id": "84c6d80c-3d4a-4772-856c-0b3825bc5333",
            "imageUrl": "http://thisImage.jpeg",
            "name": "Malvaceae",
            "price": 400,
            "totalPrice": 2000,
            "units": 2,
            "userId": "2794fddd-14f3-4ec9-a1dc-88ad2bc649da",
          }
        }
        removeFromCart={[Function]}
        showMore={[Function]}
      />
    </div>
  </div>
  <div />
</div>
`;
